daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala

```diff
@@ -29,11 +53,81 @@ object TransactionTraversal {
+      nodes: Map[String, Node],
+      toVisit: FrontStack[(RawTransaction.NodeId, Set[Ref.Party])],
+      packagesToParties: Map[String, Set[Ref.Party]] = Map.empty,
+  ): Either[ConversionError, Map[String, Set[Ref.Party]]] = {
+    toVisit match {
+      case FrontStack() => Right(packagesToParties)
+      case FrontStackCons((nodeId, parentWitnesses), toVisit) =>
+        val node = nodes(nodeId.value)
+        informeesOfNode(txVersion, node) match {
+          case Left(error) => Left(ConversionError.DecodeError(error))
+          case Right(nodeWitnesses) =>
+            val witnesses = parentWitnesses union nodeWitnesses
+            def addPackage(templateId: ValueOuterClass.Identifier) = {
+              val currentNodePackagesWithWitnesses = Map(
+                templateId.getPackageId -> witnesses
+              )
+              packagesToParties |+| currentNodePackagesWithWitnesses
+            }
+            node.getNodeTypeCase match {
+              case Node.NodeTypeCase.EXERCISE =>
+                val exercise = node.getExercise
+                // Recurse into children (if any).
+                val next = exercise.getChildrenList.asScala.view
+                  .map(RawTransaction.NodeId(_) -> witnesses)
+                  .to(ImmArray)
+                val currentNodePackagesWithWitnesses =
+                  Map(exercise.getTemplateId.getPackageId -> witnesses) ++
+                    Option
+                      .when(exercise.hasInterfaceId)(
+                        exercise.getInterfaceId.getPackageId -> witnesses
+                      )
+                      .toList
+                      .toMap
+                traverseWitnessesWithPackages(
+                  txVersion,
+                  nodes,
+                  next ++: toVisit,
+                  packagesToParties |+| currentNodePackagesWithWitnesses,
+                )
+              case Node.NodeTypeCase.FETCH =>
+                traverseWitnessesWithPackages(
+                  txVersion,
+                  nodes,
+                  toVisit,
+                  addPackage(node.getFetch.getTemplateId),
+                )
+              case Node.NodeTypeCase.CREATE =>
+                traverseWitnessesWithPackages(
+                  txVersion,
+                  nodes,
+                  toVisit,
+                  addPackage(node.getCreate.getTemplateId),
+                )
+              case Node.NodeTypeCase.LOOKUP_BY_KEY =>
+                traverseWitnessesWithPackages(
+                  txVersion,
+                  nodes,
+                  toVisit,
+                  addPackage(node.getLookupByKey.getTemplateId),
+                )
+              case Node.NodeTypeCase.NODETYPE_NOT_SET | Node.NodeTypeCase.ROLLBACK =>
+                traverseWitnessesWithPackages(txVersion, nodes, toVisit, packagesToParties)
+            }
+        }
+    }
+  }
 
   @tailrec
-  private def go(
+  private def traverseWitnesses(
```

Comment:
Why `traverseWitness` has not been treated in the same way `traverseWitnessesWithPackages` has been ?